apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,438 @@
+   */
+  _cacheGetAdapterPromise: null,
+
+  /**
+   * Store a reference of the Bluetooth API.
+   *
+   * @private
+   */
+  _bluetooth: null,
+
+  /**
+   * Store a reference of the Bluetooth API.
+   *
+   * @private
+   */
+  _isEnabled: false,
+
+  /**
+   * Build-in Bluetooth profiles
+   *
+   * @public
+   */
+  get Profiles() {
+    return {
+      HFP: 'hfp',   // Hands-Free Profile
+      OPP: 'opp',   // Object Push Profile
+      A2DP: 'a2dp', // A2DP status
+      SCO: 'sco'    // Synchronous Connection-Oriented
+    };
+  },
+
+  /**
+   * Set profile connect state.
+   *
+   * @public
+   * @param {String} profile profile id
+   * @param {Boolean} connected connect status
+   */
+  _setProfileConnected: function bt_setProfileConnected(profile, connected) {
+    var value = this['_' + profile + 'Connected'];
+    if (value !== connected) {
+      this['_' + profile + 'Connected'] = connected;
+
+      // Raise an event for the profile connection changes.
+      window.dispatchEvent(new CustomEvent('bluetoothprofileconnectionchange',
+      {
+        name: profile,
+        connected: connected
+      }));
+      // if (profile === 'opp' && this.transferIcon) {
+      //   this.transferIcon.update();
+      // }
+      // if (profile === 'a2dp' && this.headphoneIcon) {
+      //   this.headphoneIcon.update();
+      // }
+    }
+  },
+
+  /**
+   * Check if bluetooth profile is connected.
+   *
+   * @public
+   * @param {String} profile profile name
+   * @return {Boolean} connected state
+   */
+  isProfileConnected: function bt_isProfileConnected(profile) {
+    var isConnected = this['_' + profile + 'Connected'];
+    if (isConnected === undefined) {
+      return false;
+    } else {
+      return isConnected;
+    }
+  },
+
+  /**
+   * Initialize bluetooth module.
+   *
+   * @public
+   */
+  start: function bt_start() {
+    if (!window.navigator.mozSettings || !window.navigator.mozBluetooth) {
+      return;
+    }
+
+    this._bluetooth = window.navigator.mozBluetooth;
+    SettingsListener.observe('bluetooth.enabled', true, (value) => {
+      if (!this._bluetooth) {
+        // roll back the setting value to notify the UIs
+        // that Bluetooth interface is not available
+        if (value) {
+          navigator.mozSettings.createLock()
+            .set({'bluetooth.enabled': false});
+        }
+        return;
+      }
+    });
+
+    this._defaultAdapterChangeHandler =
+      this._bluetoothManagerHandler.bind(this, null);
+    this._bluetooth.addEventListener('attributechanged',
+      this._defaultAdapterChangeHandler);
+
+    // clear defaultAdapter and listener once adapter is removed
+    this._bluetooth.addEventListener('adapterremoved',
+      this._removeEventListeners.bind(this));
+
+    // if bluetooth is enabled in booting time, try to get adapter now
+    this._debug('init bluetooth adapter');
+    this._adapter = this._bluetooth.defaultAdapter;
+    if (this._adapter) {
```

Comment:
generally system bt always trapped into the above path (adapter is ready) and three cases are handled by:

1. null -> adapter: handled by this._bluetooth.addEventListener('attributechanged
2. adapter -> null: handled by this._bluetooth.addEventListener('adapterremoved'
3. adapterA -> adapterB: handled by  getAdapter

I'm not sure what is missing here